modify VIPServer in order to take into account the start_time for Merge

anchenchai opened the pull request:
This commit aims at resolving the problem if there is a "BEGIN" message sended by VIPServer but the Merge process is still waiting(before Start_time).

I remove the Timer process which was in case "GATE_PROCESS" because we don't need this any more after the modif.

I add a Timer process in case "GATE_DISCONNECT". The functionality that i want to realize is to make VIPServer wait before Merge process starts. And in order to be able to manage more than one Merge jobs(maybe useful in future), i change the "if" to "while" in line 129.  The second "while"(line 130- 137) is used to suspend VIPServer if no merge is ready to run. And i also add "				runningMergeWorkers--" when a Merge job disconnect.

frs69wq commented on src/VIPServer.java:
```diff
@@ -107,24 +107,6 @@ public void main(String[] args) throws HostFailureException, HostNotFoundExcepti
 					// SHOULD BE REPLACED BY
 					// job.carryOn();
 				} else {
-					if (!timer) {
```
You should not remove this block. Check with Sorina but it corresponds to what is done by the real system. It aims at avoiding to wait for stragglers among worker nodes and trigger the merge even if some workers are not done.

camarasu replied:
I confirm this corresponds to the real "dynamic" workflow. However, what may be a bit confusing here is that, if job.carryOn() is not active, we actually simulate the "static" workflow, for which there is no such timer. For the static workflow we wait for all jobs to finish before launching the merge (except for a very special case that we haven't yet included in the simulator). But I agree that the timer should not be completely removed since we will need it when job.carryOn() will be activated. 
To avoid confusion, we could add a few comments explaining (i) what version is being simulated (dynamic VS static) and (ii) what should be (de)activated when we switch versions. Or even better, we should implement the 2 cases distinctively.

frs69wq replied:
Most of the code has been written with the dynamic mode in mind (as I thought it was the best and most used version). For instance, in the static mode, we should not care about counting the number of particles either, but we do.
Distinguishing the modes would imply to:
1) add a new flag on the simulator command line
2) add a switch in the CARRY_ON of Gate.Java to either compute for a duration (dynamic) or a number of particles (static). In the static mode, the job never sends PROGRESS message but directly goes to the "END" section to send a DISCONNECT message. This way, we may not have to modify the code of VIPServer, or only slightly.
     Note: I'm not sure a GATE job has a number of particles as input right now ... Does this mean we have to divide the total number by the number of jobs?

camarasu replied:
Counting the number of particles in the static mode is not a problem (we actually do it in the real workflow for information purposes). 

I agree with the changes required to distinguish the 2. 

Regarding the last question, since in simulation we actually inject execution times, the number of particles is not really important. So I don't think we need to bother with getting the total number of particles and dividing it among workers. The way it is currently implemented seems fine to me.

anchenchai replied:
So to conclude, I will add a new flag on the simulator command line to distinguish the 2 modes (Static VS Dynamic).
 1) In static,  Gate jobs directly enter END after computation and a Timer in VIPServer to start merge( as I have added in GATE_DISCONNECT)
  2) In dynamic, Gate jobs send PROGRESS and I add back the block that i removed in GATE_PROGRESS ?
    Is that correct and you all agree with these modifs?

frs69wq replied:
So if in VIP we send back the number of computing particles to the server, we have to do it in simulation too. 
For the last point the idea is to stop injecting the execution time at some point, isn't it? So my question is: in VIP how is distributed the initial number of particles among the workers? we have to do the same way, even if we do not use it for now. 

My point is that it's better to be as close as possible to the internal logic of VIP from the beginning than having to patch up things later because suddenly there is a "but is not like that in real life".

frs69wq replied:
About point 1) maybe a unique sleep from the time the GATE job ends (obtained with getClock) and the time at which the Merge job starts would be cleaner. In fact I'm not even sure we need something here. when GATE jobs end they upload their files and are stopped. The merge starts when it has to and reads what's available. What was the motivation for a timer here?

anchenchai replied:
Ok, about point 1), if we implement like this, it will work. Using getClock to compute the delay between GATE job ends and Merge job starts. But in this case, we imply that the VIPServer knows how long the queuing time of Merge job is. I feel it is a little less realist. 

Maybe it is the name of process "Timer" makes confusion. In fact, the reason why i created a new process in GATE_DISCONNECT is to test whether Merge job really starts, if not, VIPServer waits for a retry delay and test again because if we dont do this test, a "BEGIN" may be sended to Merge job before merge process really starts, which may cause errors (a msg to a process not existing).
In this implementation, the VIPServer does not know the exact queuing time of Merge job. And in order to not block the main process VIPServer, i create a new process.

frs69wq replied:
What is in GATE_DISCONNECT assumed that the Merge process was launched at the beginning of the simulation and waits to be activated. If the process is launched at its start_time, they are 2 options:
1) it is before the last GATE_DISCONNECT and the current code handles this case
2) it is after the last GATE_DISCONNECT and then we have to test in MERGE_CONNECT if gate jobs are done and if yes, start the process:
if (endedGateWorkers == VIPSimulator.numberOfGateJobs &&  
    runningMergeWorkers < VIPSimulator.numberOfMergeJobs) {
  runningMergeWorkers++;
   job.begin();
}
otherwise, we do nothing more than now. But there is no need for polling from the server.

camarasu replied:
In reply to Fred's question on "in VIP how is distributed the initial number of particles among the workers?", the short answer is: "for the static workflow, each node computes the same fraction of the work". The long answer is that the fraction is computed either on the number of particles, or on the GATE time runs (which are very different of the real time the execution takes). Moreover, this is not decided at the workflow level (VIPServer), but at the worker level, based on the parsing of the GATE macro file. If we want to go to this level of details, there will be a few things we need to add to the simulator and the log parser (to detect whether the splitting was done on time or events). 

So I do believe that the details of the splitting algorithm can be left aside for a while, as long as we manage to reproduce the functioning by injecting the execution time.

Yes, we will stop injecting the execution time at some point, but this is a long-term goal. In my opinion, we should handle the splitting at the time we will really need it. It is not about "patching up things later", is about adding elements step by step.

frs69wq replied:
not so long term ... before the end of the PhD would be great ;) 
I'm ok to skip it for now, but it bothers me to have a half-baked support of computing time in the code.

frs69wq commented on src/VIPServer.java:
```diff
@@ -139,10 +121,27 @@ public void main(String[] args) throws HostFailureException {
+											runningMergeWorkers++;
+											mergeFlag = true;
+										}
+										else Process.sleep(10);
```
10 milliseconds are too short, increase this retry delay

frs69wq commented on src/VIPServer.java:
```diff
@@ -139,10 +121,27 @@ public void main(String[] args) throws HostFailureException {
+								Process.sleep(VIPSimulator.sosTime);
+								boolean mergeFlag = false;
+								while (runningMergeWorkers < VIPSimulator.numberOfMergeJobs) {							
+									while(!mergeFlag){
```
Please use more expressive variable names

frs69wq commented on src/VIPServer.java:
```diff
@@ -139,10 +121,27 @@ public void main(String[] args) throws HostFailureException {
+								while (runningMergeWorkers < VIPSimulator.numberOfMergeJobs) {							
+									while(!mergeFlag){
+										if(mergeWorkers.size() > 0){
+											mergeWorkers.get(runningMergeWorkers).begin();
```
This will start the first merger in the vector (possibly several times), not all of them

frs69wq commented on src/VIPServer.java:
```diff
@@ -139,10 +121,27 @@ public void main(String[] args) throws HostFailureException {
+										if(mergeWorkers.size() > 0){
+											mergeWorkers.get(runningMergeWorkers).begin();
+											runningMergeWorkers++;
+											mergeFlag = true;
```
yopur flag is not reset to false in this outer while. This can only start a single merger

anchenchai commented:
I reverted last commit.
In the new commit, I have added the new flag for workflow version (Static and Dynamic) and modified the code to cleanly evoke Merge job without polling server.

frs69wq commented on src/VIPSimulator.java:
```diff
@@ -51,6 +52,8 @@ public static void main(String[] args) {
 		Msg.info("PARAMS:   sostime is " + sosTime + ", number of Gate tasks is " + numberOfGateJobs
 				+ ", number of merge tasks is " + numberOfMergeJobs + ", cpu merge time is " + cpuMergeTime);
 
+		workflowVersion = args.length > 10 ? args[10] : "dynamic";
```
maybe use "static" as a default value here, as we cannot fully handle the dynamic case right now. Moreover, with a "static" default we don't have to modify the launchers.

anchenchai commented:
Besides 10 arguments for simulators, there are still 2 flags for simgrid. So need to test "args.length > 12" if we want to set "workflowVersion" to "static"  by default without modify the launcher.
